### PR TITLE
Fix compilation warnings

### PR DESCRIPTION
**Pull request by gonzaponte**
Fixes compilation warnings in gcc. Some variables have been renamed to avoid shadowing.

One option to address #6.

### CODE REVIEW

**Comment by jacg** [comment]
The semicolons in the first commit have to go, I completely agree.

As for the second commit, I fundamentally disagree with the compiler: This kind of shadowing *in constructors* is a common C++ idiom, and mangling the names to avoid the shadowing is a complete waste of time and space.

Looks like I'm not the only one: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=78147. I completely agree with comment 7:

> I think that GCC should _never_ warn about this usage.  I really can't fathom a reason that anyone would want to enable shadow warnings in this situation

So, no, I'm not going to budge on this. It's a bug in the compiler. They should fix it.

In the meantime, what can we do to mitigate this deluge of noise? 

From the bug report, I infer that `gcc` doesn't distinguish between shadowing in constructors (where it is C++ idiom) and elsewhere (where it *might* cause confusion), so if we use [gcc diagnostic pragmas](https://gcc.gnu.org/onlinedocs/gcc/Diagnostic-Pragmas.html), they would have to be pushed and popped in a fairly fine-grained way. Nah, I don't want all that noise in our code. So I propose that we just disable the shadow warning once per file in *our* code.

**Comment by gonzaponte** [comment]
I was also able to silent those warnings by removing the `-Wshadow` flag in `CMakeLists.txt` if you want to check this option. I didn't suggest it because I wasn't comfortable with disabling it _everywhere_. Do you want me to submit this option?

**Comment by jacg** [comment]
You were right to be uncomfortable: we should not be disabling the shadowing warning everywhere; only in that particular usage in the constructors, where it has no business being on in the first place.

**Comment by jacg** [comment]
Cachix binary cache seems to be working. I've populated it for Linux from my own machine, but we have to wait about an hour until the MacOS Geant4 is built on GHA and pushed to the cache. Then we can run the GHA tests with a turnaround of a few minutes, instead of the hour it used to take to recompile G4 each time. Then I can unleash all the blocked GHA runs on your many PRs.

File: nain4/nain4.hh [hunk 1]
```diff
@@ -33,6 +33,14 @@
 #include <optional>
 
 
+// Disable GCC's shadow warnings
+// This happens in constructors where it is a C++ idiom to use
+// the same name for the argument and the class member.
```
Comment (jacg) [inline nain4/nain4.hh:38]: Just a bit of my pet pedantry: **arguments** are *values* that are passed in to functions; the *names* to which those values are bound when they arrive inside the function are **parameters**.

About 90% of the time, we don't care about the difference, and we can conflate the two terminologies. But sometimes the difference *does* matter, and then it's useful to have the linguistic wherewithal to distinguish them, which is why I insist on using the correct terminology everywhere.

In this example you could even argue that your use of the term 'argument' is correct, because a parameter *is* a 'name for the argument', but let's just use 'parameter' anyway. 

TLDR: 'it is an idiom to have constructor parameters match member names'

Comment (gonzaponte) [reply nain4/nain4.hh:38]: It's good that you keep pinpointing this because I keep forgetting :)

File: nain4/nain4.hh [hunk 2]
```diff
@@ -34,8 +34,8 @@
 
 
 // Disable GCC's shadow warnings
-// This happens in constructors where it is a C++ idiom to use
-// the same name for the argument and the class member.
+// This happens in constructors where it is a C++ idiom to have
+// constructor parameters match memeber names'.
```
Comment (jacg) [inline nain4/nain4.hh:38]: Stray quote/apostrophe at the end?

Nothing belongs to the names, so it's not possessive, and I can't see an opening quote.

**Comment by gonzaponte** [comment]
Do you understand the build failure? Seems like it's not linking to G4.

**Comment by jacg** [comment]
No, I don't understand it. But as I said in #15, we have to forget MacOS for a while, and make some progress on other things. It doesn't make sense to waste more time on MacOS, when we don't have access to a Mac and while tons of high-priority work is being delayed by it.
